<a href="https://colab.research.google.com/github/hrtnisri2016/vix-ds-hci/blob/main/Home_Credit_Default_Risk_(Data_Validation).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


**Import library**

In [ ]:
import os
import textwrap
import pandas as pd
import numpy as np
from IPython.display import display, HTML

**Some useful functions**

In [ ]:
# function to create scrollable table within a small window
def create_scrollable_table(df, table_id, title):
    html = f'<h3>{title}</h3>'
    html += f'<div id="{table_id}" style="height:250px; overflow:auto">'
    html += df.to_html(classes="scrollable-table")
    html += '</div>'
    return html

In [ ]:
# function to create scrollable output within a small window
def create_scrollable_output(content, output_id, title):
    html = f'<h3>{title}</h3>'
    html += f'<pre id="{output_id}" style="height:200px; overflow:auto;">'
    html += content
    html += '</pre>'
    return html

**Get data**

In [ ]:
folder_path = '/content/drive/MyDrive/home-credit-default-risk'

file_dict = {}  # Dictionary to store file paths

# List all files in the folder
for root, dirs, files in os.walk(folder_path):
    for file in files:
        file_path = os.path.join(root, file)
        file_name = os.path.splitext(file)[0]  # Get the file name without extension
        file_dict[file_name] = file_path

# Print the dictionary
for file_name, file_path in file_dict.items():
    print(f'\u2022 {file_name}\n  {file_path}')

• HomeCredit_columns_description
  /content/drive/MyDrive/home-credit-default-risk/HomeCredit_columns_description.csv
• POS_CASH_balance
  /content/drive/MyDrive/home-credit-default-risk/POS_CASH_balance.csv
• application_test
  /content/drive/MyDrive/home-credit-default-risk/application_test.csv
• application_train
  /content/drive/MyDrive/home-credit-default-risk/application_train.csv
• bureau
  /content/drive/MyDrive/home-credit-default-risk/bureau.csv
• bureau_balance
  /content/drive/MyDrive/home-credit-default-risk/bureau_balance.csv
• credit_card_balance
  /content/drive/MyDrive/home-credit-default-risk/credit_card_balance.csv
• installments_payments
  /content/drive/MyDrive/home-credit-default-risk/installments_payments.csv
• previous_application
  /content/drive/MyDrive/home-credit-default-risk/previous_application.csv
• sample_submission
  /content/drive/MyDrive/home-credit-default-risk/sample_submission.csv


### **Sample submission**

The sample submission should consist of just two columns: 
*   `SK_ID_CURR`: the loan ID in our sample and
*   `TARGET`: the predicted credit score.

In [ ]:
path_sample_submission = file_dict['sample_submission']

df_sample_submission = pd.read_csv(path_sample_submission)
print(f'shape: {df_sample_submission.shape}')
df_sample_submission.head()

shape: (48744, 2)


SK_ID_CURR  TARGET
0      100001     0.5
1      100005     0.5
2      100013     0.5
3      100028     0.5
4      100038     0.5

### **Description of Home Credit data columns**

In [ ]:
path_HomeCredit_columns_description = file_dict['HomeCredit_columns_description']

df_HomeCredit_columns_description = pd.read_csv(path_HomeCredit_columns_description, encoding='latin-1')
df_HomeCredit_columns_description.fillna('', inplace=True)
print(f'shape: {df_HomeCredit_columns_description.shape}')

out_HomeCredit_columns_description = create_scrollable_table(
    df_HomeCredit_columns_description, "HomeCredit_columns_description", "HomeCredit_columns_description"
)
display(HTML(out_HomeCredit_columns_description))

shape: (219, 5)


### **Application table: train and test dataset**

In [ ]:
# Load application_train data
path_application_train = file_dict['application_train']

df_application_train = pd.read_csv(path_application_train)
print(f'shape: {df_application_train.shape}')
df_application_train.head()

shape: (307511, 122)


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 122 columns]

In [ ]:
df_application_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 286.2+ MB


In [ ]:
# Load application_test data
path_application_test = file_dict['application_test']

df_application_test = pd.read_csv(path_application_test)
print(f'shape: {df_application_test.shape}')
df_application_test.info()

shape: (48744, 121)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48744 entries, 0 to 48743
Columns: 121 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(40), object(16)
memory usage: 45.0+ MB


Both `application_train` and `application_test` datasets have the same data types for their features. They consist of 65 columns with `float64` data type, 40 columns with `int64` data type, and 16 columns with `object` data type.

In the `application_train` dataset, there is an additional column with an `int64` data type, which is the `TARGET` column. This column does not exist in the `application_test` dataset.

In [ ]:
# Select columns of 'object' data type in the application_train DataFrame
object_columns = df_application_train.select_dtypes(include='object').columns

# Iterate over each object column
output = ''
for column in object_columns:
    # Get the unique values in the column, excluding NaN values
    unique_values_without_nan = df_application_train[column].dropna().unique()
    # Format the unique values as a comma-separated string with line wrapping
    formatted_values = ', '.join(map(str, unique_values_without_nan))
    formatted_values = textwrap.fill(formatted_values, width=120, subsequent_indent=' '*8)
    # Print the column name and its unique values
    output += f'\u2022 {column}:\n  {formatted_values}\n'

content = create_scrollable_output(output, "unique_values", "Here are the unique values present in each column")
display(HTML(content))


If you carefully examine every unique value in the dataset, you will notice the presence of `'XNA'`, `'Unknown'`, and `'not specified'`. These values indicate that certain information was not provided by the user. To ensure data consistency, we can convert these values to `NaN` (null values) so they are recognized as missing values.

In [ ]:
# Replace 'XNA', 'Unknown', and 'not specified' with NaN
df_application_train.replace(['XNA', 'Unknown', 'not specified'], np.nan, inplace=True)
df_application_test.replace(['XNA', 'Unknown', 'not specified'], np.nan, inplace=True)

To ensure data integrity, it is also advisable to check if there are any columns that contain only a single unique value. This can help identify columns that do not provide useful information for analysis.

In [ ]:
# Iterate over each column
for column in df_application_train.columns:
    # Get the unique non-null values in the column
    unique_values = df_application_train[column].dropna().nunique()
    
    # Check if there is more than one unique non-null value
    if unique_values <= 1:
        print(f"Column '{column}' consists of only one unique non-null value.")

No column consists of only one value.

In [ ]:
# Generate summary statistics for the application_train dataset and round to 1 decimal places
summary_statistics_train = df_application_train.describe().T.round(1)

out_summary_statistics_train = create_scrollable_table(
    summary_statistics_train, "summary_statistics_train", "Summary statistics")
display(HTML(out_summary_statistics_train))

In [ ]:
summary_statistics_train[summary_statistics_train['min'] < 0]

count     mean       std      min      25%  \
DAYS_BIRTH              307511.0 -16037.0    4364.0 -25229.0 -19682.0   
DAYS_EMPLOYED           307511.0  63815.0  141275.8 -17912.0  -2760.0   
DAYS_REGISTRATION       307511.0  -4986.1    3522.9 -24672.0  -7479.5   
DAYS_ID_PUBLISH         307511.0  -2994.2    1509.5  -7197.0  -4299.0   
DAYS_LAST_PHONE_CHANGE  307510.0   -962.9     826.8  -4292.0  -1570.0   

                            50%      75%       max  
DAYS_BIRTH             -15750.0 -12413.0   -7489.0  
DAYS_EMPLOYED           -1213.0   -289.0  365243.0  
DAYS_REGISTRATION       -4504.0  -2010.0       0.0  
DAYS_ID_PUBLISH         -3254.0  -1720.0       0.0  
DAYS_LAST_PHONE_CHANGE   -757.0   -274.0       0.0

It's worth noting that the `max` value of 365,243 in the `DAYS_EMPLOYED` column is an anomaly in the dataset. This value, which corresponds to approximately 1,000 years, is often used as a placeholder to indicate missing or unknown  information. 

In [ ]:
# Generate summary statistics for the application_test dataset and round to 1 decimal places
summary_statistics_test = df_application_test.describe().T.round(1)
summary_statistics_test[summary_statistics_test['min'] < 0]

count     mean       std      min      25%  \
DAYS_BIRTH                   48744.0 -16068.1    4325.9 -25195.0 -19637.0   
DAYS_EMPLOYED                48744.0  67485.4  144348.5 -17463.0  -2910.0   
DAYS_REGISTRATION            48744.0  -4967.7    3552.6 -23722.0  -7459.2   
DAYS_ID_PUBLISH              48744.0  -3051.7    1569.3  -6348.0  -4448.0   
REGION_RATING_CLIENT_W_CITY  48744.0      2.0       0.5     -1.0      2.0   
DAYS_LAST_PHONE_CHANGE       48744.0  -1077.8     878.9  -4361.0  -1766.2   

                                 50%      75%       max  
DAYS_BIRTH                  -15785.0 -12496.0   -7338.0  
DAYS_EMPLOYED                -1293.0   -296.0  365243.0  
DAYS_REGISTRATION            -4490.0  -1901.0       0.0  
DAYS_ID_PUBLISH              -3234.0  -1706.0       0.0  
REGION_RATING_CLIENT_W_CITY      2.0      2.0       3.0  
DAYS_LAST_PHONE_CHANGE        -863.0   -363.0       0.0

It appears that there is an error in the input of data for the `REGION_RATING_CLIENT_W_CITY` column. The expected values should be ratings 1, 2, or 3, but there is a value of -1 present. Assuming this is a data input mistake, we can correct it by replacing it with a value of 1.

In [ ]:
# Replace 365243 with NaN in DAYS_EMPLOYED column
df_application_train['DAYS_EMPLOYED'].replace(365243, np.nan, inplace=True)
df_application_test['DAYS_EMPLOYED'].replace(365243, np.nan, inplace=True)
# Replace -1 with 1 in REGION_RATING_CLIENT_W_CITY column 
df_application_test['REGION_RATING_CLIENT_W_CITY'].replace(-1, 1, inplace=True)

The maximum value of `AMT_INCOME_TOTAL` is 117,000,000.0. 

When we look at individuals with `AMT_INCOME_TOTAL` over a million, the average value doesn't seem suspicious, except for a specific group consisting of "Working" in the `NAME_INCOME_TYPE` and "Laborers" in the `OCCUPATION_TYPE`. In this group, the average `AMT_INCOME_TOTAL` is unusually high, exceeding 9 million, while other groups have averages ranging from 1 to 3 million. 

What's even stranger is that within the Laborers group, we find the individual with an `AMT_INCOME_TOTAL` of 117,000,000.0, and their `TARGET` value is 1, indicating payment difficulties. This appears to be another data input error, and although I can't verify it directly, I will remove this record from the dataset to ensure it doesn't affect further analysis.

In [ ]:
suspect_df = df_application_train[df_application_train['AMT_INCOME_TOTAL'] >= 1000000][
    ['TARGET', 'NAME_CONTRACT_TYPE', 'AMT_INCOME_TOTAL', 'DAYS_EMPLOYED', 'NAME_INCOME_TYPE', 'OCCUPATION_TYPE']]
suspect_df.groupby(['NAME_INCOME_TYPE', 'OCCUPATION_TYPE'])['AMT_INCOME_TOTAL'].mean().round()

NAME_INCOME_TYPE      OCCUPATION_TYPE      
Businessman           Managers                 2250000.0
Commercial associate  Accountants              1416537.0
                      Core staff               1228221.0
                      Drivers                  1350000.0
                      High skill tech staff    1395000.0
                      IT staff                 2025000.0
                      Laborers                 1373674.0
                      Managers                 1432414.0
                      Private service staff    1192500.0
                      Sales staff              1253571.0
                      Secretaries              1800000.0
State servant         Core staff               1770750.0
                      Managers                 1305000.0
                      Medicine staff           1260000.0
                      Private service staff    3150000.0
Working               Accountants              1237500.0
                      Cleaning staff        

In [ ]:
df_laborers = suspect_df[suspect_df.OCCUPATION_TYPE == 'Laborers']
laborers = create_scrollable_table(df_laborers, "laborers", "Laborers")
display(HTML(laborers))

In [ ]:
# Exclude observations with AMT_INCOME_TOTAL equal to 117000000 from df_application_train
df_application_train = df_application_train.query('AMT_INCOME_TOTAL != 117000000')

In [ ]:
# Calculate null proportions for every column
null_proportions = df_application_train.isnull().mean() * 100
null_proportions_df = null_proportions.to_frame(name='Null_Proportions')

out_null_proportion = create_scrollable_table(
    null_proportions_df.query('Null_Proportions != 0').sort_values('Null_Proportions', ascending=False), 'null_proportion', '')

display(HTML(out_null_proportion))

In [ ]:
df_application_train[df_application_train['OWN_CAR_AGE'].isnull()]['FLAG_OWN_CAR'].value_counts()

N    202923
Y         5
Name: FLAG_OWN_CAR, dtype: int64

The `OWN_CAR_AGE` column contains approximately 66% null values. However, upon closer examination, we found that among those null values, 202,923 observations belong to individuals who do not own a car (`FLAG_OWN_CAR = 'N'`). Therefore, for these specific rows, it is more appropriate to fill the null values with 0.

In [ ]:
# Fill null values in OWN_CAR_AGE column based on FLAG_OWN_CAR

# Get indices where FLAG_OWN_CAR is 'N' and OWN_CAR_AGE is null
indices = df_application_train[
    (df_application_train['FLAG_OWN_CAR'] == 'N') & df_application_train['OWN_CAR_AGE'].isnull()].index

# Fill null values with 0 for the selected indices
df_application_train.loc[indices, 'OWN_CAR_AGE'] = 0

We have found that several columns contain more than 47% null values. The proportion of null values in these columns ranges from approximately 47% to 70%. To clean the data, we will exclude these columns from our dataset, except for the `EXT_SOURCE_1` column, which may be useful for future analysis.

In [ ]:
# Calculate null value proportions for each column
null_proportions = df_application_train.drop('EXT_SOURCE_1', axis=1).isnull().mean()

# Identify columns with null proportions greater than 47% (excluding EXT_SOURCE_1)
columns_to_drop = null_proportions[null_proportions > 0.47].index.tolist()
print(f'Delete {len(columns_to_drop)} columns.')

# Drop the identified columns from the DataFrame
df_application_train.drop(columns_to_drop, axis=1, inplace=True)
print(f'Current shape: {df_application_train.shape}')

Delete 47 columns.
Current shape: (307510, 75)


### **Bureau**

In [ ]:
# Load bureau data
path_bureau = file_dict['bureau']

df_bureau = pd.read_csv(path_bureau)
print(f'shape: {df_bureau.shape}\n')
df_bureau.head()

shape: (1716428, 17)



SK_ID_CURR  SK_ID_BUREAU CREDIT_ACTIVE CREDIT_CURRENCY  DAYS_CREDIT  \
0      215354       5714462        Closed      currency 1         -497   
1      215354       5714463        Active      currency 1         -208   
2      215354       5714464        Active      currency 1         -203   
3      215354       5714465        Active      currency 1         -203   
4      215354       5714466        Active      currency 1         -629   

   CREDIT_DAY_OVERDUE  DAYS_CREDIT_ENDDATE  DAYS_ENDDATE_FACT  \
0                   0               -153.0             -153.0   
1                   0               1075.0                NaN   
2                   0                528.0                NaN   
3                   0                  NaN                NaN   
4                   0               1197.0                NaN   

   AMT_CREDIT_MAX_OVERDUE  CNT_CREDIT_PROLONG  AMT_CREDIT_SUM  \
0                     NaN                   0         91323.0   
1                     NaN                   0        225000.0   
2                     NaN                   0        464323.5   
3                     NaN                   0         90000.0   
4                 77674.5                   0       2700000.0   

   AMT_CREDIT_SUM_DEBT  AMT_CREDIT_SUM_LIMIT  AMT_CREDIT_SUM_OVERDUE  \
0                  0.0                   NaN                     0.0   
1             171342.0                   NaN                     0.0   
2                  NaN                   NaN                     0.0   
3                  NaN                   NaN                     0.0   
4                  NaN                   NaN                     0.0   

       CREDIT_TYPE  DAYS_CREDIT_UPDATE  AMT_ANNUITY  
0  Consumer credit                -131          NaN  
1      Credit card                 -20          NaN  
2  Consumer credit                 -16          NaN  
3      Credit card                 -16          NaN  
4  Consumer credit                 -21          NaN

In [ ]:
df_bureau.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1716428 entries, 0 to 1716427
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   SK_ID_CURR              int64  
 1   SK_ID_BUREAU            int64  
 2   CREDIT_ACTIVE           object 
 3   CREDIT_CURRENCY         object 
 4   DAYS_CREDIT             int64  
 5   CREDIT_DAY_OVERDUE      int64  
 6   DAYS_CREDIT_ENDDATE     float64
 7   DAYS_ENDDATE_FACT       float64
 8   AMT_CREDIT_MAX_OVERDUE  float64
 9   CNT_CREDIT_PROLONG      int64  
 10  AMT_CREDIT_SUM          float64
 11  AMT_CREDIT_SUM_DEBT     float64
 12  AMT_CREDIT_SUM_LIMIT    float64
 13  AMT_CREDIT_SUM_OVERDUE  float64
 14  CREDIT_TYPE             object 
 15  DAYS_CREDIT_UPDATE      int64  
 16  AMT_ANNUITY             float64
dtypes: float64(8), int64(6), object(3)
memory usage: 222.6+ MB


In [ ]:
# Select columns of 'object' data type in the bureau DataFrame
object_columns = df_bureau.select_dtypes(include='object').columns

# Iterate over each object column
output = ''
for column in object_columns:
    # Get the unique values in the column
    unique_values = df_bureau[column].unique()
    # Format the unique values as a comma-separated string with line wrapping
    formatted_values = ', '.join(map(str, unique_values))
    formatted_values = textwrap.fill(formatted_values, width=90, subsequent_indent=' '*8)
    # Print the column name and its unique values
    output += f'\u2022 {column}:\n  {formatted_values}\n'

content = create_scrollable_output(output, "unique_values", "Here are the unique values present in each column")
display(HTML(content))

There are two categories in the CREDIT_TYPE column: "Unknown type of loan" and "Another type of loan." These categories represent loans that are either unknown or don't fit into the pre-defined types. To simplify and group them together, I will combine them into a single category called "Unknown type of loan."

In [ ]:
# Combine "Unknown type of loan" and "Another type of loan" into "Unknown type of loan"
df_bureau['CREDIT_TYPE'] = df_bureau['CREDIT_TYPE'].replace(['Unknown type of loan', 'Another type of loan'], 
                                                            'Unknown type of loan')

# Print the updated unique values in the CREDIT_TYPE column
print(df_bureau['CREDIT_TYPE'].unique())

['Consumer credit' 'Credit card' 'Mortgage' 'Car loan' 'Microloan'
 'Loan for working capital replenishment' 'Loan for business development'
 'Real estate loan' 'Unknown type of loan' 'Cash loan (non-earmarked)'
 'Loan for the purchase of equipment' 'Mobile operator loan'
 'Interbank credit' 'Loan for purchase of shares (margin lending)']


In [ ]:
# Generate summary statistics for the bureau dataset and round to 1 decimal places
summary_statistics_bureau = df_bureau.describe().T.round(1)

out_summary_statistics_bureau = create_scrollable_table(
    summary_statistics_bureau, "summary_statistics_bureau", "Summary statistics")
display(HTML(out_summary_statistics_bureau))

In [ ]:
# Calculate null proportions for every column
null_proportions = df_bureau.isnull().mean() * 100
null_proportions_df = null_proportions.to_frame(name='Null_Proportions')

out_null_proportion = create_scrollable_table(
    null_proportions_df.query('Null_Proportions != 0').sort_values('Null_Proportions', ascending=False), 'null_proportion', '')

display(HTML(out_null_proportion))

In [ ]:
# Identify columns with null proportions greater than 50%
columns_to_drop = null_proportions[null_proportions > 50].index.tolist()
print(f'Delete {len(columns_to_drop)} columns.')

# Drop the identified columns from the DataFrame
df_bureau.drop(columns_to_drop, axis=1, inplace=True)
print(f'Current shape: {df_bureau.shape}')

Delete 2 columns.
Current shape: (1716428, 15)


### **Bureau balance**

In [ ]:
# Load bureau_balance data
path_bureau_balance = file_dict['bureau_balance']

df_bureau_balance = pd.read_csv(path_bureau_balance)
print(f'shape: {df_bureau_balance.shape}\n')
df_bureau_balance.head()

shape: (27299925, 3)



SK_ID_BUREAU  MONTHS_BALANCE STATUS
0       5715448               0      C
1       5715448              -1      C
2       5715448              -2      C
3       5715448              -3      C
4       5715448              -4      C

In [ ]:
# Print unique values from STATUS column
df_bureau_balance['STATUS'].unique()

array(['C', '0', 'X', '1', '2', '3', '5', '4'], dtype=object)

In [ ]:
# view summary statistics of numeric features
df_bureau_balance.describe().T.round(1)

count       mean       std        min        25%  \
SK_ID_BUREAU    27299925.0  6036297.3  492348.9  5001709.0  5730933.0   
MONTHS_BALANCE  27299925.0      -30.7      23.9      -96.0      -46.0   

                      50%        75%        max  
SK_ID_BUREAU    6070821.0  6431951.0  6842888.0  
MONTHS_BALANCE      -25.0      -11.0        0.0

In [ ]:
# Calculate the number of null values
df_bureau_balance.isnull().sum()

SK_ID_BUREAU      0
MONTHS_BALANCE    0
STATUS            0
dtype: int64

### **POS (point-of-sale) transactions and cash balance**

In [ ]:
# Load POS_CASH_balance data
path_POS_CASH_balance = file_dict['POS_CASH_balance']

df_POS_CASH_balance = pd.read_csv(path_POS_CASH_balance)
print(f'shape: {df_POS_CASH_balance.shape}\n')
df_POS_CASH_balance.head()

shape: (10001358, 8)



SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  CNT_INSTALMENT  \
0     1803195      182943             -31            48.0   
1     1715348      367990             -33            36.0   
2     1784872      397406             -32            12.0   
3     1903291      269225             -35            48.0   
4     2341044      334279             -35            36.0   

   CNT_INSTALMENT_FUTURE NAME_CONTRACT_STATUS  SK_DPD  SK_DPD_DEF  
0                   45.0               Active       0           0  
1                   35.0               Active       0           0  
2                    9.0               Active       0           0  
3                   42.0               Active       0           0  
4                   35.0               Active       0           0

In [ ]:
# Print unique values from NAME_CONTRACT_STATUS column
df_POS_CASH_balance['NAME_CONTRACT_STATUS'].unique()

array(['Active', 'Completed', 'Signed', 'Approved',
       'Returned to the store', 'Demand', 'Canceled', 'XNA',
       'Amortized debt'], dtype=object)

Once again, we have encountered 'XNA', so we will replace it with null values.

In [ ]:
# Replace 'XNA' with NaN
df_POS_CASH_balance.replace('XNA', np.nan, inplace=True)

In [ ]:
# view summary statistics of numeric features
df_POS_CASH_balance.describe().T.round(1)

count       mean       std        min        25%  \
SK_ID_PREV             10001358.0  1903216.6  535846.5  1000001.0  1434405.0   
SK_ID_CURR             10001358.0   278403.9  102763.7   100001.0   189550.0   
MONTHS_BALANCE         10001358.0      -35.0      26.1      -96.0      -54.0   
CNT_INSTALMENT          9975287.0       17.1      12.0        1.0       10.0   
CNT_INSTALMENT_FUTURE   9975271.0       10.5      11.1        0.0        3.0   
SK_DPD                 10001358.0       11.6     132.7        0.0        0.0   
SK_DPD_DEF             10001358.0        0.7      32.8        0.0        0.0   

                             50%        75%        max  
SK_ID_PREV             1896565.0  2368963.0  2843499.0  
SK_ID_CURR              278654.0   367429.0   456255.0  
MONTHS_BALANCE             -28.0      -13.0       -1.0  
CNT_INSTALMENT              12.0       24.0       92.0  
CNT_INSTALMENT_FUTURE        7.0       14.0       85.0  
SK_DPD                       0.0        0.0     4231.0  
SK_DPD_DEF                   0.0        0.0     3595.0

In [ ]:
# Calculate null proportions for every column
null_proportions = df_POS_CASH_balance.isnull().mean() * 100
null_proportions_df = null_proportions.to_frame(name='Null_Proportions')
null_proportions_df[null_proportions_df['Null_Proportions'] > 0]

Null_Proportions
CNT_INSTALMENT                 0.260675
CNT_INSTALMENT_FUTURE          0.260835
NAME_CONTRACT_STATUS           0.000020

### Credit card balance

In [ ]:
# Load credit_card_balance data
path_credit_card_balance = file_dict['credit_card_balance']

df_credit_card_balance = pd.read_csv(path_credit_card_balance)
print(f'shape: {df_credit_card_balance.shape}\n')
df_credit_card_balance.head()

shape: (3840312, 23)



SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  AMT_BALANCE  \
0     2562384      378907              -6       56.970   
1     2582071      363914              -1    63975.555   
2     1740877      371185              -7    31815.225   
3     1389973      337855              -4   236572.110   
4     1891521      126868              -1   453919.455   

   AMT_CREDIT_LIMIT_ACTUAL  AMT_DRAWINGS_ATM_CURRENT  AMT_DRAWINGS_CURRENT  \
0                   135000                       0.0                 877.5   
1                    45000                    2250.0                2250.0   
2                   450000                       0.0                   0.0   
3                   225000                    2250.0                2250.0   
4                   450000                       0.0               11547.0   

   AMT_DRAWINGS_OTHER_CURRENT  AMT_DRAWINGS_POS_CURRENT  \
0                         0.0                     877.5   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                   11547.0   

   AMT_INST_MIN_REGULARITY  ...  AMT_RECIVABLE  AMT_TOTAL_RECEIVABLE  \
0                 1700.325  ...          0.000                 0.000   
1                 2250.000  ...      64875.555             64875.555   
2                 2250.000  ...      31460.085             31460.085   
3                11795.760  ...     233048.970            233048.970   
4                22924.890  ...     453919.455            453919.455   

   CNT_DRAWINGS_ATM_CURRENT  CNT_DRAWINGS_CURRENT  CNT_DRAWINGS_OTHER_CURRENT  \
0                       0.0                     1                         0.0   
1                       1.0                     1                         0.0   
2                       0.0                     0                         0.0   
3                       1.0                     1                         0.0   
4                       0.0                     1                         0.0   

   CNT_DRAWINGS_POS_CURRENT  CNT_INSTALMENT_MATURE_CUM  NAME_CONTRACT_STATUS  \
0                       1.0                       35.0                Active   
1                       0.0                       69.0                Active   
2                       0.0                       30.0                Active   
3                       0.0                       10.0                Active   
4                       1.0                      101.0                Active   

   SK_DPD  SK_DPD_DEF  
0       0           0  
1       0           0  
2       0           0  
3       0           0  
4       0           0  

[5 rows x 23 columns]

In [ ]:
# Print unique values from NAME_CONTRACT_STATUS column
df_credit_card_balance['NAME_CONTRACT_STATUS'].unique()

array(['Active', 'Completed', 'Demand', 'Signed', 'Sent proposal',
       'Refused', 'Approved'], dtype=object)

In [ ]:
# view summary statistics of numeric features
desc_CC_balance = df_credit_card_balance.describe().T.round(1)
out_desc_CC_balance = create_scrollable_table(
    desc_CC_balance, "summary_statistics_CC", "Summary statistics"
)
display(HTML(out_desc_CC_balance))

In [ ]:
# Calculate null proportions for every column
null_proportions = df_credit_card_balance.isnull().mean() * 100
null_proportions_df = null_proportions.to_frame(name='Null_Proportions')

out_null_proportion = create_scrollable_table(
    null_proportions_df.query('Null_Proportions != 0').sort_values('Null_Proportions', ascending=False), 'null_proportion', '')

display(HTML(out_null_proportion))

### **Previous application**

In [ ]:
# Load previous_application data
path_previous_application = file_dict['previous_application']

df_previous_application = pd.read_csv(path_previous_application)
print(f'shape: {df_previous_application.shape}\n')
df_previous_application.head()

shape: (1670214, 37)



SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START  ... NAME_SELLER_INDUSTRY  CNT_PAYMENT  \
0                       15  ...         Connectivity         12.0   
1                       11  ...                  XNA         36.0   
2                       11  ...                  XNA         12.0   
3                        7  ...                  XNA         12.0   
4                        9  ...                  XNA         24.0   

   NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0            middle  POS mobile with interest            365243.0   
1        low_action          Cash X-Sell: low            365243.0   
2              high         Cash X-Sell: high            365243.0   
3            middle       Cash X-Sell: middle            365243.0   
4              high         Cash Street: high                 NaN   

  DAYS_FIRST_DUE DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
0          -42.0                     300.0          -42.0            -37.0   
1         -134.0                     916.0       365243.0         365243.0   
2         -271.0                      59.0       365243.0         365243.0   
3         -482.0                    -152.0         -182.0           -177.0   
4            NaN                       NaN            NaN              NaN   

  NFLAG_INSURED_ON_APPROVAL  
0                       0.0  
1                       1.0  
2                       1.0  
3                       1.0  
4                       NaN  

[5 rows x 37 columns]

In [ ]:
# Select columns of 'object' data type in the previous_application DataFrame
object_columns = df_previous_application.select_dtypes(include='object').columns

# Iterate over each object column
output = ''
for column in object_columns:
    # Get the unique values in the column
    unique_values = df_previous_application[column].unique()
    # Format the unique values as a comma-separated string with line wrapping
    formatted_values = ', '.join(map(str, unique_values))
    formatted_values = textwrap.fill(formatted_values, width=90, subsequent_indent=' '*8)
    # Print the column name and its unique values
    output += f'\u2022 {column}:\n  {formatted_values}\n'

content = create_scrollable_output(output, "unique_values", "Here are the unique values present in each column")
display(HTML(content))

We have noticed multiple occurrences of `XNA` in various columns, including 

* `NAME_CONTRACT_TYPE`, 
* `NAME_CASH_LOAN_PURPOSE`, 
* `NAME_PAYMENT_TYPE`, 
* `CODE_REJECT_REASON`, 
* `NAME_CLIENT_TYPE`, 
* `NAME_GOODS_CATEGORY`, 
* `NAME_PORTFOLIO`, 
* `NAME_PRODUCT_TYPE`, 
* `NAME_SELLER_INDUSTRY`, and 
* `NAME_YIELD_GROUP`. 

We will replace these instances with null values. 

In the `NAME_CASH_LOAN_PURPOSE` column, we will also replace `XAP` and `Refusal to name the goal` with null values. However, `XAP` in the `CODE_REJECT_REASON` column will not be changed to null as it represents a special category indicating that the application was canceled or withdrawn by the applicant before the lender made a decision.

In [ ]:
# Replace 'XNA' with NaN
df_previous_application.replace('XNA', np.nan, inplace=True)

# Replace "XAP" and "Refusal to name the goal" with null values in the NAME_CASH_LOAN_PURPOSE column
replace_values = ['XAP', 'Refusal to name the goal']
df_previous_application['NAME_CASH_LOAN_PURPOSE'] = df_previous_application['NAME_CASH_LOAN_PURPOSE']\
                                                        .replace(replace_values, np.nan)

In [ ]:
# view summary statistics of numeric features
desc_prev_app = df_previous_application.describe().T.round(1)
out_desc_prev_app = create_scrollable_table(
    desc_prev_app, "summary_statistics_prev_app", "Summary statistics"
)
display(HTML(out_desc_prev_app))

In [ ]:
df_previous_application.query('SELLERPLACE_AREA < 0')['SELLERPLACE_AREA'].value_counts()

-1    762675
Name: SELLERPLACE_AREA, dtype: int64

The value of -1 is used as a default value for missing or unavailable `SELLERPLACE_AREA` information. However, it should be replaced with `NaN`, which represents the absence of data.

In [ ]:
# Replace the default value of -1 with NaN
df_previous_application['SELLERPLACE_AREA'].replace(-1, np.nan, inplace=True)

The value of 365,243 in the columns 
* `DAYS_FIRST_DRAWING`, 
* `DAYS_FIRST_DUE`, 
* `DAYS_LAST_DUE_1ST_VERSION`, 
* `DAYS_LAST_DUE`, and 
* `DAYS_TERMINATION` 

is commonly used as a placeholder to represent missing or unknown information. 

It should be replaced with `NaN` to accurately indicate the absence of data.

In [ ]:
# Replace the placeholder value of 365,243 with NaN
columns_to_replace = [
    'DAYS_FIRST_DRAWING', 
    'DAYS_FIRST_DUE', 
    'DAYS_LAST_DUE_1ST_VERSION', 
    'DAYS_LAST_DUE', 
    'DAYS_TERMINATION'
]
df_previous_application[columns_to_replace] = df_previous_application[columns_to_replace].replace(365243, np.nan)

In [ ]:
# Calculate null proportions for every column
null_proportions = df_previous_application.isnull().mean() * 100
null_proportions_df = null_proportions.to_frame(name='Null_Proportions')

out_null_proportion = create_scrollable_table(
    null_proportions_df.query('Null_Proportions != 0').sort_values('Null_Proportions', ascending=False), 'null_proportion', '')

display(HTML(out_null_proportion))

In [ ]:
# Identify columns with null proportions greater than 50%
columns_to_drop = null_proportions[null_proportions > 40].index.tolist()
print(f'Delete {len(columns_to_drop)} columns.')

# Drop the identified columns from the DataFrame
df_previous_application.drop(columns_to_drop, axis=1, inplace=True)
print(f'Current shape: {df_previous_application.shape}')

Delete 16 columns.
Current shape: (1670214, 21)


### **Installments payments**

In [ ]:
# Load installments_payments data
path_installments_payments = file_dict['installments_payments']

df_installments_payments = pd.read_csv(path_installments_payments)
print(f'shape: {df_installments_payments.shape}\n')
df_installments_payments.head()

shape: (13605401, 8)



SK_ID_PREV  SK_ID_CURR  NUM_INSTALMENT_VERSION  NUM_INSTALMENT_NUMBER  \
0     1054186      161674                     1.0                      6   
1     1330831      151639                     0.0                     34   
2     2085231      193053                     2.0                      1   
3     2452527      199697                     1.0                      3   
4     2714724      167756                     1.0                      2   

   DAYS_INSTALMENT  DAYS_ENTRY_PAYMENT  AMT_INSTALMENT  AMT_PAYMENT  
0          -1180.0             -1187.0        6948.360     6948.360  
1          -2156.0             -2156.0        1716.525     1716.525  
2            -63.0               -63.0       25425.000    25425.000  
3          -2418.0             -2426.0       24350.130    24350.130  
4          -1383.0             -1366.0        2165.040     2160.585

In [ ]:
# view summary statistics
df_installments_payments.describe().T.round(1)

count       mean       std        min        25%  \
SK_ID_PREV              13605401.0  1903365.0  536202.9  1000001.0  1434191.0   
SK_ID_CURR              13605401.0   278444.9  102718.3   100001.0   189639.0   
NUM_INSTALMENT_VERSION  13605401.0        0.9       1.0        0.0        0.0   
NUM_INSTALMENT_NUMBER   13605401.0       18.9      26.7        1.0        4.0   
DAYS_INSTALMENT         13605401.0    -1042.3     800.9    -2922.0    -1654.0   
DAYS_ENTRY_PAYMENT      13602496.0    -1051.1     800.6    -4921.0    -1662.0   
AMT_INSTALMENT          13605401.0    17050.9   50570.3        0.0     4226.1   
AMT_PAYMENT             13602496.0    17238.2   54735.8        0.0     3398.3   

                              50%        75%        max  
SK_ID_PREV              1896520.0  2369094.0  2843499.0  
SK_ID_CURR               278685.0   367530.0   456255.0  
NUM_INSTALMENT_VERSION        1.0        1.0      178.0  
NUM_INSTALMENT_NUMBER         8.0       19.0      277.0  
DAYS_INSTALMENT            -818.0     -361.0       -1.0  
DAYS_ENTRY_PAYMENT         -827.0     -370.0       -1.0  
AMT_INSTALMENT             8884.1    16710.2  3771487.8  
AMT_PAYMENT                8125.5    16108.4  3771487.8

In [ ]:
# Calculate null proportions for every column
null_proportions = df_installments_payments.isnull().mean() * 100
null_proportions_df = null_proportions.to_frame(name='Null_Proportions')
null_proportions_df[null_proportions_df['Null_Proportions'] != 0]

Null_Proportions
DAYS_ENTRY_PAYMENT          0.021352
AMT_PAYMENT                 0.021352

### Save the newly validated dataset.

In [ ]:
# Define the main path where the files will be saved
main_path = '/content/drive/MyDrive/home-credit-default-risk'

# Save every DataFrame to a CSV file without index
df_application_train.to_csv(f'{main_path}/application_train_valid.csv', index=False)
df_application_test.to_csv(f'{main_path}/application_test_valid.csv', index=False)
df_bureau.to_csv(f'{main_path}/bureau_valid.csv', index=False)
df_POS_CASH_balance.to_csv(f'{main_path}/POS_CASH_balance_valid.csv', index=False)
df_previous_application.to_csv(f'{main_path}/previous_application_valid.csv', index=False)